File: python_bindings_jupedsim/routing.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
#include "CollisionGeometry.hpp"
#include "Floorfield.hpp"
#include "RoutingEngine.hpp"
#include "conversion.hpp"

#include <glm/ext/vector_float2.hpp>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h> // IWYU pragma: keep

#include <cstddef>
#include <cstdint>
#include <memory>
#include <tuple>
#include <vector>

namespace py = pybind11;

void init_routing(py::module_& m)
{
    using Floorfield = ::Floorfield<>; // concrete instantiation with default Scalar
    py::class_<RoutingEngine>(m, "RoutingEngine")
        .def(py::init([](const CollisionGeometry& geo) {
            return std::make_unique<RoutingEngine>(geo.Polygon());
        }))
        .def(
            "compute_waypoints",
            [](RoutingEngine& engine,
               std::tuple<double, double> from,
               std::tuple<double, double> to) {
                return intoTuples(engine.ComputeAllWaypoints(intoPoint(from), intoPoint(to)));
            })
        .def(
            "is_routable",
            [](RoutingEngine& engine, std::tuple<double, double> point) {
                return engine.IsRoutable(intoPoint(point));
            })
        .def("mesh", [](const RoutingEngine& routingEngine) {
            const auto mesh = routingEngine.MeshData();
            const auto polygonCount = mesh->CountPolygons();
            using Ind = decltype(mesh->Polygons(0).vertices);
            std::vector<Ind> polys;
            polys.reserve(polygonCount);
            for(size_t index = 0; index < polygonCount; ++index) {
                const auto& poly = mesh->Polygons(index);
                polys.emplace_back(poly.vertices);
            }
            return std::make_tuple(intoTuples(mesh->FVertices()), polys);
        });

    py::class_<Floorfield>(m, "Floorfield")
        .def(
            py::init([](const CollisionGeometry& geo, double wallInfluenceRadius) {
                return std::make_unique<Floorfield>(geo.Polygon(), wallInfluenceRadius);
            }),
            py::arg("geometry"),
            py::arg("wall_influence_radius") = 0.5)
        .def(
            "compute_waypoints",
            [](Floorfield& ff, std::tuple<double, double> from, std::tuple<double, double> to) {
                return intoTuples(ff.ComputeAllWaypoints(intoPoint(from), intoPoint(to)));
            })
        .def(
            "is_routable",
            [](Floorfield& ff, std::tuple<double, double> point) {
                return ff.IsRoutable(intoPoint(point));
            })
        .def(
            "speed_field",
            [](const Floorfield& ff) {
                // Returns a dict ready for numpy: reshape data as
                // np.array(d["data"]).reshape(d["height"], d["width"])
                // Extent for imshow: [origin_x, origin_x + width*cell_size,
                //                      origin_y, origin_y + height*cell_size]
                py::dict d;
                d["width"] = ff.GridWidth();
                d["height"] = ff.GridHeight();
                d["origin"] = std::make_tuple(ff.Origin().x, ff.Origin().y);
                d["cell_size"] = ff.CellSize();
                d["data"] = ff.SpeedField();
                return d;
            })
        .def(
            "travel_times",
            [](const Floorfield& ff) {
                // Same layout as speed_field; call compute_waypoints first to
                // populate the field for a specific destination.
                py::dict d;
                d["width"] = ff.GridWidth();
                d["height"] = ff.GridHeight();
                d["origin"] = std::make_tuple(ff.Origin().x, ff.Origin().y);
                d["cell_size"] = ff.CellSize();
                d["data"] = ff.TravelTimes();
                return d;
            })
        .def(
            "set_eikonal_solver",
            [](Floorfield& ff, const std::string& name) {
                if(name == "FSM")
                    ff.SetSolver(EikonalSolver::FSM);
                else if(name == "FMM")
                    ff.SetSolver(EikonalSolver::FMM);
                else if(name == "FIM")
                    ff.SetSolver(EikonalSolver::FIM);
                else
                    throw std::invalid_argument("Unknown solver: " + name + " (use FSM, FMM, FIM)");
            },
            py::arg("solver"),
            "Select the eikonal solver: 'FSM', 'FMM', or 'FIM'. Disables auto-benchmark.")
        .def(
            "clear_point_cache",
            [](Floorfield& ff) { ff.ClearPointCache(); },
            "Clear the point-destination floor-field cache and reset the step counter.")
        .def(
            "update_density",
            [](Floorfield& ff, std::vector<std::tuple<double, double>> positions) {
                std::vector<Point> pts;
                pts.reserve(positions.size());
                for(const auto& p : positions)
                    pts.push_back(intoPoint(p));
                ff.UpdateDensity(std::span<const Point>(pts));
            },
            py::arg("positions"),
            "Update the density field from agent positions. Rebuilds dynamic speed field every "
            "set_recompute_interval() calls.")
        .def(
            "density_field",
            [](const Floorfield& ff) {
                py::dict d;
                d["width"] = ff.GridWidth();
                d["height"] = ff.GridHeight();
                d["origin"] = std::make_tuple(ff.Origin().x, ff.Origin().y);
                d["cell_size"] = ff.CellSize();
                d["data"] = ff.DensityField();
                return d;
            },
            "Returns raw agent density (agents/m²) per cell. Same dict layout as speed_field().")
        .def(
            "dynamic_speed_field",
            [](const Floorfield& ff) {
                py::dict d;
                d["width"] = ff.GridWidth();
                d["height"] = ff.GridHeight();
                d["origin"] = std::make_tuple(ff.Origin().x, ff.Origin().y);
                d["cell_size"] = ff.CellSize();
                d["data"] = ff.DynamicSpeedField();
                return d;
            },
            "Returns density-modulated speed field (0–1) per cell. Call update_density() first.")
        .def(
            "set_recompute_interval",
            [](Floorfield& ff, int steps) { ff.SetRecomputeInterval(steps); },
            py::arg("steps"),
            "Rebuild density/dynamic-speed fields every *steps* calls to update_density().")
        .def(
            "precompute_destinations",
            [](Floorfield& ff, std::vector<std::tuple<double, double>> points) {
                std::vector<Point> pts;
                pts.reserve(points.size());
                for(const auto& p : points)
                    pts.push_back(intoPoint(p));
                std::vector<size_t> ids;
                ff.PrecomputeDestinations(
                    std::span<const size_t>(ids), std::span<const Point>(pts));
            },
            py::arg("points"),
            "Precompute floor fields for the given list of (x, y) point destinations.");
}

File: python_modules/jupedsim/jupedsim/routing.py
# SPDX-License-Identifier: LGPL-3.0-or-later

from typing import Any

import shapely

import jupedsim.native as py_jps
from jupedsim.geometry_utils import build_geometry

_GridData = dict  # {"width": int, "height": int, "origin": tuple, "cell_size": float, "data": list[float]}


class RoutingEngine:
    """RoutingEngine to compute the shortest paths with navigation meshes."""

    def __init__(
        self,
        geometry: (
            str
            | shapely.GeometryCollection
            | shapely.Polygon
            | shapely.MultiPolygon
            | shapely.MultiPoint
            | list[tuple[float, float]]
        ),
        **kwargs: Any,
    ) -> None:
        self._obj = py_jps.RoutingEngine(
            build_geometry(geometry, **kwargs)._obj
        )

    def compute_waypoints(
        self, frm: tuple[float, float], to: tuple[float, float]
    ) -> list[tuple[float, float]]:
        """Computes shortest path between specified points.

        Arguments:
            geometry: Data to create the geometry out of. Data may be supplied as:

                * list of 2d points describing the outer boundary, holes may be added with use of `excluded_areas` kw-argument

                * :class:`~shapely.GeometryCollection` consisting only out of :class:`Polygons <shapely.Polygon>`, :class:`MultiPolygons <shapely.MultiPolygon>` and :class:`MultiPoints <shapely.MultiPoint>`

                * :class:`~shapely.MultiPolygon`

                * :class:`~shapely.Polygon`

                * :class:`~shapely.MultiPoint` forming a "simple" polygon when points are interpreted as linear ring without repetition of the start/end point.

                * str with a valid Well Known Text. In this format the same WKT types as mentioned for the shapely types are supported: GEOMETRYCOLLETION, MULTIPOLYGON, POLYGON, MULTIPOINT. The same restrictions as mentioned for the shapely types apply.

            frm: point from which to find the shortest path
            to: point to which to find the shortest path

        Keyword Arguments:
            excluded_areas: describes exclusions
                from the walkable area. Only use this argument if `geometry` was
                provided as list[tuple[float, float]].

        Returns:
            List of points (path) from 'frm' to 'to' including from and to.

        """
        return self._obj.compute_waypoints(frm, to)

    def is_routable(self, p: tuple[float, float]) -> bool:
        """Tests if the supplied point is inside the underlying geometry.

        Returns:
            If the point is inside the geometry.

        """
        return self._obj.is_routable(p)

    def mesh(
        self,
    ) -> tuple[list[tuple[float, float]], list[list[int]]]:
        """Access the navigation mesh geometry.

        The navigation mesh is store as a collection of convex polygons in CCW order.

        The returned data is to be interpreted as:

        .. code::

            tuple[
                list[tuple[float, float]], # All vertices in this mesh.
                list[ # List of polygons
                    list[int] # List of indices into the vertices that compose this polygon in CCW order
                ]
            ]

        Returns:
            A tuple of vertices and list of polygons which in turn are a list of indices
            tuple[list[tuple[float, float]],list[list[int]]]
        """
        return self._obj.mesh()

    def edges_for(self, vertex_id: int):
        return self._obj.edges_for(vertex_id)


class Floorfield:
    """Floor-field router that solves the eikonal equation on a regular grid.

    The walkable domain is rasterised from the supplied geometry at construction
    time.  Travel times are computed on demand for each queried destination and
    cached so that repeated queries for the same destination are cheap.
    """

    def __init__(
        self,
        geometry: (
            str
            | shapely.GeometryCollection
            | shapely.Polygon
            | shapely.MultiPolygon
            | shapely.MultiPoint
            | list[tuple[float, float]]
        ),
        *,
        wall_influence_radius: float = 0.5,
        **kwargs: Any,
    ) -> None:
        """Create a floor-field router for the given geometry.

        Arguments:
            geometry: walkable area — same formats as :class:`RoutingEngine`.
            wall_influence_radius: cells within this distance (metres) of any
                boundary edge receive a proportionally reduced speed, so the
                eikonal naturally routes paths away from walls.  Default 0.5 m.

        Keyword Arguments:
            excluded_areas: holes subtracted from the walkable area (only when
                *geometry* is a ``list[tuple[float, float]]``).
        """
        self._obj = py_jps.Floorfield(
            build_geometry(geometry, **kwargs)._obj, wall_influence_radius
        )

    def compute_waypoints(
        self, frm: tuple[float, float], to: tuple[float, float]
    ) -> list[tuple[float, float]]:
        """Compute the path from *frm* to *to* via gradient descent on the floor field.

        The floor field for *to* is computed on the first call and cached.

        Arguments:
            frm: start point ``(x, y)``
            to:  destination point ``(x, y)``

        Returns:
            Ordered list of waypoints from *frm* to *to*, inclusive.
        """
        return self._obj.compute_waypoints(frm, to)

    def is_routable(self, p: tuple[float, float]) -> bool:
        """Return ``True`` if *p* lies inside the walkable domain."""
        return self._obj.is_routable(p)

    def speed_field(self) -> _GridData:
        """Return the walkability grid built from the geometry.

        The returned dict has the keys:

        * ``"width"``     – number of columns
        * ``"height"``    – number of rows
        * ``"origin"``    – ``(x, y)`` of the bottom-left corner of cell (0, 0)
        * ``"cell_size"`` – physical size of one cell in metres
        * ``"data"``      – flat row-major list; ``1.0`` = walkable, ``0.0`` = obstacle

        Reconstruct a 2-D array and an imshow extent with::

            import numpy as np
            d = ff.speed_field()
            grid = np.array(d["data"]).reshape(d["height"], d["width"])
            ox, oy = d["origin"]
            cs = d["cell_size"]
            extent = [ox, ox + d["width"] * cs, oy, oy + d["height"] * cs]
        """
        return self._obj.speed_field()

    def travel_times(self) -> _GridData:
        """Return the travel-time grid for the most recently queried destination.

        Has the same layout as :meth:`speed_field`.  Call
        :meth:`compute_waypoints` at least once before using this so the
        eikonal solution is populated.  Cells unreachable from the destination
        carry ``float("inf")``.
        """
        return self._obj.travel_times()

    def update_density(self, positions: list[tuple[float, float]]) -> None:
        """Update the density field from agent positions.

        Rebuilds the dynamic speed field every :meth:`set_recompute_interval`
        calls.  Has no effect until the rebuild threshold is reached.

        Arguments:
            positions: list of ``(x, y)`` agent positions.
        """
        self._obj.update_density(positions)

    def density_field(self) -> _GridData:
        """Return raw agent density (agents/m²) per cell.

        Has the same layout as :meth:`speed_field`.  Call
        :meth:`update_density` first to populate.
        """
        return self._obj.density_field()

    def dynamic_speed_field(self) -> _GridData:
        """Return the density-modulated speed field (0–1) per cell.

        Has the same layout as :meth:`speed_field`.  Values drop below the
        static :meth:`speed_field` in high-density cells.  Call
        :meth:`update_density` first to populate.
        """
        return self._obj.dynamic_speed_field()

    def set_recompute_interval(self, steps: int) -> None:
        """Set how many :meth:`update_density` calls trigger a full rebuild.

        A value of ``1`` rebuilds on every call; higher values amortise the
        cost across multiple frames at the expense of staleness.

        Arguments:
            steps: number of calls between rebuilds (default 200).
        """
        self._obj.set_recompute_interval(steps)
